In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!unrar e /content/drive/'My Drive'/DataMining/dataset.rar


UNRAR 5.50 freeware      Copyright (c) 1993-2017 Alexander Roshal


Extracting from /content/drive/My Drive/DataMining/dataset.rar

Extracting  product.xlsx                                                   5% 10% 16% 17%  OK 
Extracting  tarikhche kharid.csv                                          22% 27% 33% 38% 43% 48%  OK 
Extracting  comment.xlsx                                                  53% 58% 64% 69%  OK 
Extracting  keifiat.xlsx                                                  69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94%  OK 
Extracting  orders.csv                                                    99%  OK 
All OK


In [ ]:
import pandas as pd 
import numpy as np
import datetime
import re
print('done')

done


# **Orders dataset**

In [ ]:
df_orders = pd.read_csv('orders.csv')
df_orders

ID_Order  ID_Customer  ...  city_name_fa Quantity_item
0        2714054       469662  ...    محمود آباد           1.0
1       11104039      3063877  ...        خرمدره           1.0
2        4228130      3184893  ...          قرچک           1.0
3       22225624      6888562  ...            قم           1.0
4        4068771      2533490  ...         تهران           1.0
...          ...          ...  ...           ...           ...
199995  12172771      4403268  ...         تهران           2.0
199996   4425393      3745774  ...        محمدیه           1.0
199997   6671889      4845514  ...        اصفهان           1.0
199998   6315995      2541418  ...         همدان           1.0
199999   7584127      1342585  ...    بندر انزلی           1.0

[200000 rows x 7 columns]

In [ ]:
e = df_orders.loc[1, ['DateTime_CartFinalize']]
print(type(e))

<class 'pandas.core.series.Series'>


\\
with this line we check if the time is in the future compare to now. This shows all of the rows of the ```DateTime_CartFinalize``` column are valid beacuse, there is no record that the belongs to a future time.

In [ ]:
df_orders[(pd.to_datetime(df_orders['DateTime_CartFinalize']) > datetime.datetime.now())]

Empty DataFrame
Columns: [ID_Order, ID_Customer, ID_Item, DateTime_CartFinalize, Amount_Gross_Order, city_name_fa, Quantity_item]
Index: []

\\
These two lines below checks shows that all of the rows of ```Amount_Gross_Order``` and ```Quantity_item``` columns are integer

In [ ]:
all(x.is_integer() for x in df_orders['Amount_Gross_Order'])

True

In [ ]:
all(x.is_integer() for x in df_orders['Quantity_item'])

True

\\
Code below shows what is the range of the id of the orders in which we can see that the all the ids consist of 5 digits.

In [ ]:
minimum = min(df_orders['ID_Order'])
maximum = max(df_orders['ID_Order'])
print('minimum = ', minimum)
print('maximum = ', maximum)

minimum =  1000411
maximum =  24846558


\\
# **Product dataset**

In [ ]:
df_product = pd.read_excel('product.xlsx')
df_product

id  ...                                 product_attributes
0      723302  ...  [{"Key":"وزن بسته‌بندی","Value":"480"},{"Key":...
1      727089  ...  [{"Key":"دستگیره","Value":"1 جفت"},{"Key":"نوع...
2      727096  ...  [{"Key":"حجم","Value":"50"},{"Key":"نوع عصاره"...
3      727803  ...  [{"Key":"ابعاد رو بالشی","Value":"50×70"},{"Ke...
4      725222  ...  [{"Key":"در","Value":"ندارد"},{"Key":"جنس","Va...
...       ...  ...                                                ...
99995  803375  ...                                                NaN
99996  790932  ...                                                NaN
99997  786604  ...                                                NaN
99998  799047  ...                                                NaN
99999  782276  ...                                                NaN

[100000 rows x 10 columns]

# **Comments dataset**

In [ ]:
df_comment = pd.read_excel('comment.xlsx')
df_comment

product_id  ...                                            comment
0            90213  ...  سلام، قبل اینکه نظرم رو بگم میخواستم به یک موض...
1            56871  ...  بوی تند ولی خوشبو داره.ماندگاریش خوبه و هر چقد...
2            49738  ...  متاسفانه عمر مفید این ماشین کم هست و بعد از دو...
3           134367  ...  افا از وقتی من اینو واسه پسرم خریدم همه بچه ها...
4           151402  ...  این تیغ محصول کشور آمریکا هست ( در عکس محصول م...
...            ...  ...                                                ...
100003      104998  ...  قیمت مناسبی داره در حال حاضر . با این پول سخت ...
100004      333389  ...  در کل راضیم درسته لنوو با قیمت پایین تر هسته و...
100005      356546  ...  من این محصول رو از بیرون تهیه کردم . و باید بگ...
100006       33874  ...  دوسال و.نیم پیش از دیجی کالا گرفتم. هنوز مثل ر...
100007      318278  ...  با استفاده از این شامپو مشکل خشکی مو و پوست سر...

[100008 rows x 3 columns]

In [ ]:
df_comment.loc[(pd.to_datetime(df_comment['confirmed_at']) > datetime.datetime.now())]

Empty DataFrame
Columns: [product_id, confirmed_at, comment]
Index: []

The code below shows that all of the our dataset product_id have an integer value.

In [ ]:
pattern = re.compile(r'[a-zA-Z]*')
df_keifiat[~(df_keifiat.product_id.apply(lambda x: bool(pattern.match(str(x).strip()))))]

Empty DataFrame
Columns: [product_id, product_title, title_en, user_id, likes, dislikes, verification_status, recommend, title, comment, advantages, disadvantages]
Index: []

\\
Code below shows what is the range of the id of the orders. As can be seen, the minimum value for a product_id is 281 and this is a bit suspicious.

In [ ]:
minimum = min(df_comment['product_id'])
maximum = max(df_comment['product_id'])
print('minimum = ', minimum)
print('maximum = ', maximum)

minimum =  281
maximum =  684859


# **keifiat dataset**

In [ ]:
df_keifiat = pd.read_excel('keifiat.xlsx')
df_keifiat

product_id  ...                                      disadvantages
0            3692  ...                                                NaN
1           90213  ...                                          ["ندارد"]
2           59473  ...                                                NaN
3          120499  ...  ["کیفیت پایین اسپیکر\r","حاشیه خالی زیر صفحه ن...
4           67200  ...                                    ["ظرفیت متوسط"]
...           ...  ...                                                ...
99995      549845  ...                                          ["ندارد"]
99996      784160  ...                 ["محل عبور بند کفش ها ایراد دارد"]
99997      820985  ...                                          ["ندیدم"]
99998      678073  ...                                                NaN
99999      451791  ...                                                NaN

[100000 rows x 12 columns]

\\
The ```verification_status``` has only consisted of verified, rejected, not_verified.

*   verified: 98496
*   rejected: 1263
*   not_verified: 241


In [ ]:
df_keifiat[(df_keifiat['verification_status'] != 'verified') & (df_keifiat['verification_status'] != 'rejected')]

product_id  ...                     disadvantages
94         183430  ...  ["گارانتی، در خدمات پس از فروش"]
149        147179  ...                               NaN
258        203451  ...                          ["هیچی"]
262         90894  ...                                []
350        263989  ...                          ["قیمت"]
...           ...  ...                               ...
96985      434472  ...                               NaN
97140      728278  ...                                []
97202      893102  ...             ["محصول فیک و تقلبی"]
97546      574637  ...                               NaN
97602      382126  ...                               NaN

[241 rows x 12 columns]

\\
**id validation**:

check which of the product_ids are not valid.
This shows that product_id of all rows only consist of numbers. As can be seen, all the product_ids are valid.

In [ ]:
pattern = re.compile(r'[a-zA-Z]*')
df_keifiat[~(df_keifiat.product_id.apply(lambda x: bool(pattern.match(str(x).strip()))))]

Empty DataFrame
Columns: [product_id, product_title, title_en, user_id, likes, dislikes, verification_status, recommend, title, comment, advantages, disadvantages]
Index: []

# **Tarikhche kharid dataset**

In [ ]:
df_tarikhche = pd.read_csv('tarikhche kharid.csv')
df_tarikhche

id  product_variant_id  ...  product_id  marketplace_seller_id
0        18752043             1977385  ...      667969                   2962
1        18752022             1977380  ...      150144                  78525
2        18752046             1977380  ...      150144                  78525
3        18751982             1977374  ...      908541                   4199
4        18752018             1977374  ...      908541                   4199
...           ...                 ...  ...         ...                    ...
2184353  12763117             1155042  ...      221109                   7664
2184354  13703581             1155042  ...      221109                   7664
2184355  13703591             1155042  ...      221109                   7664
2184356  13703595             1155042  ...      221109                   7664
2184357  16766304             1155042  ...      221109                   7664

[2184358 rows x 15 columns]

\\
Code below shows that non of the data rows have invalid create_at date.

In [ ]:
df_tarikhche[(pd.to_datetime(df_tarikhche['created_at']) > datetime.datetime.now())]

Empty DataFrame
Columns: [id, product_variant_id, selling_price, rrp_price, base_price, buy_price, order_limit, start_at, end_at, tags, show_in_price_history, active, created_at, product_id, marketplace_seller_id]
Index: []

\\
This line of code shows that, the ```active``` column has just the values of 0 or 1

In [ ]:
df_tarikhche[(df_tarikhche['active'] != 1) & (df_tarikhche['active'] != 0)]

Empty DataFrame
Columns: [id, product_variant_id, selling_price, rrp_price, base_price, buy_price, order_limit, start_at, end_at, tags, show_in_price_history, active, created_at, product_id, marketplace_seller_id]
Index: []

\\
## **not valid dataframes:**

We defined a valid price is a price which is more than 1000 so, using regex, we have filtered  the ```selling_price``` and ```rrp_price``` to find out which of the rows have a price less than 1000.

\\
### **1**

In [ ]:
lenDf1 = len(notValid_df1)

for i in range(lenDf1):
  print(notValid_df1)

725

\\

### **1**
Code below shows that, there are **725** rows which have invalid values for column ``` selling_price``` and ```rrp_price```.

In [ ]:
pattern = re.compile(r'^[0-9]{4,}')
notValid_df1 = df_tarikhche[~(df_tarikhche.selling_price.apply(lambda x: bool(pattern.match(str(x).strip())))) & 
                            ~(df_tarikhche.rrp_price.apply(lambda x: bool(pattern.match(str(x).strip()))))]
notValid_df1                        

id  product_variant_id  ...  product_id  marketplace_seller_id
20905    18274482             1926782  ...      469087                   9691
20913    18274447             1926774  ...      469087                   9691
20927    18274439             1926766  ...      469087                   9691
20931    18274545             1926766  ...      469087                   9691
20936    18274418             1926748  ...      859481                   9691
...           ...                 ...  ...         ...                    ...
2014942  17825295             1240364  ...      684376                   6167
2031706   7795384             1231314  ...      157816                  10078
2124915   7255330             1194692  ...       59785                   6085
2134552   7236502             1192407  ...      404828                  10939
2144740   7348258             1183151  ...       88874                  10939

[725 rows x 15 columns]

\\
**Strategy to fill the invalid cells:**

Replace all value in ```notValid_df1``` for columns ```selling_price``` and ```rrp_price``` to ```NaN```

\\
### **2**

Code below show that there is only 1 which has an invalid rrp_price and its rrp_price is not equal to its selling_price.

In [ ]:
pattern = re.compile(r'^[0-9]{4,}')
notValid_df2 = df_tarikhche[~(df_tarikhche.rrp_price.apply(lambda x: bool(pattern.match(str(x).strip())))) & (df_tarikhche['rrp_price'] != df_tarikhche['selling_price'])]

id  product_variant_id  ...  product_id  marketplace_seller_id
1092073  10938499             1434284  ...      679181                   8392

[1 rows x 15 columns]

\\
**Strategy to fill the invalid cells:**

Replace all value in ```notValid_df1``` for columns ```selling_price``` and ```rrp_price``` to ```NaN```

\\
### **3**

Code below show that there are 89 rows, which have an invalid selling_price and its rrp_price is not equal to its selling_price.

In [ ]:
pattern = re.compile(r'^[0-9]{4,}')
notValid_df3 = df_tarikhche[~(df_tarikhche.selling_price.apply(lambda x: bool(pattern.match(str(x).strip())))) & (df_tarikhche['rrp_price'] != df_tarikhche['selling_price'])]

id  product_variant_id  ...  product_id  marketplace_seller_id
606745   13260493             1589716  ...       23232                   8095
805911   12338768             1518976  ...      349877                   8652
982644   11338674             1464151  ...      768056                   8095
982647   11338590             1464140  ...      279167                   8095
1059900  16125301             1444315  ...      763821                      1
...           ...                 ...  ...         ...                    ...
2176033   7557728             1160964  ...      634224                     69
2176035   7576266             1160964  ...      634224                     69
2176037   7599305             1160964  ...      634224                     69
2176039   7616207             1160964  ...      634224                     69
2180806   7018650             1157096  ...       85988                  10078

[89 rows x 15 columns]

\\
**Our strategy:**

We decided to change the value of the selling_price columns to its rrp_price.